### PR TITLE
Support multiple ceph monitors

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR adds the support for multiple ceph monitors on UI and API.

image

What have been tested and work
- vm life cycle
- volume snapshot (created a separated PR #6790)
- vm migration to another host
- Stop a ceph monitor and test vm life cycle.

What have been tested but do not work (I believe they are not caused by this PR)
- vm live migration with volume to another RBD primary storage
- direct download to RBD primary storage
- VM snapshots on RBD primary storage

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

File: plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDefTest.java [hunk 1]
```diff
@@ -238,6 +238,39 @@ public void testDiskDefWithEncryption() {
         assertEquals(disk.toString(), expectedXML);
     }
 
+    @Test
+    public void testDiskDefWithMultipleHosts() {
+        String path = "/mnt/primary1";
+        String host = "10.11.12.13,10.11.12.14,10.11.12.15";
```
Comment (wido) [inline plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDefTest.java:244]: Could you also make a test-case with IPv6 addresses? Just to make sure we cover that as well

File: plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParserTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                "    </auth>\n" +
+                "  </source>\n" +
+                "</pool>";
+
+        LibvirtStoragePoolXMLParser parser = new LibvirtStoragePoolXMLParser();
+        LibvirtStoragePoolDef pool = parser.parseStoragePoolXML(poolXML);
+
+        Assert.assertEquals(LibvirtStoragePoolDef.PoolType.RBD, pool.getPoolType());
+        Assert.assertEquals(LibvirtStoragePoolDef.AuthenticationType.CEPH, pool.getAuthType());
+        Assert.assertEquals("10.11.12.13,10.11.12.14,10.11.12.15", pool.getSourceHost());
```
Comment (wido) [inline plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParserTest.java:76]: Would be good if we have an IPv6 test case here as well

File: plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParserTest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                "      <owner>0</owner>\n" +
+                "      <group>0</group>\n" +
+                "    </permissions>\n" +
+                "  </target>\n" +
+                "</pool>";
+
+        LibvirtStoragePoolXMLParser parser = new LibvirtStoragePoolXMLParser();
+        LibvirtStoragePoolDef pool = parser.parseStoragePoolXML(poolXML);
+
+        Assert.assertEquals("10.11.12.13", pool.getSourceHost());
```
Comment (wido) [inline plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParserTest.java:52]: Another IPv6 test case?

File: plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDefTest.java (hunk 1 above)
Comment (weizhouapache) [reply plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDefTest.java:244]: @wido 
It seems monitor with ipv6 address is not suported in this PR.

what is currently supported and any issue in following scenarios ?
- ipv6 addr without port
- ipv6 addr with port
I do not have a ceph env with ipv6 address , it is difficult to verify.

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParser.java [hunk 4]
```diff
@@ -123,4 +125,13 @@ private static String getAttrValue(String tag, String attr, Element eElement) {
         Element node = (Element)tagNode.item(0);
         return node.getAttribute(attr);
     }
+
+    protected static String getStorageHost(Element parentElement) {
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParser.java:129]:
```suggestion
    protected static String getStorageHosts(Element parentElement) {
```

File: plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDefTest.java (hunk 1 above)
Comment (weizhouapache) [reply plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDefTest.java:244]: @wido 
please ignore my previous comment
I will use ULA addresses in testing

Comment (wido) [reply plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDefTest.java:244]: > @wido please ignore my previous comment I will use ULA addresses in testing

You can also use the 2001:db8::/32 range for this: https://www.rfc-editor.org/rfc/rfc3849

That would be sufficient to write your test-cases with.

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParser.java (hunk 4 above)
Comment (weizhouapache) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParser.java]: done, thanks @DaanHoogland

File: plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParserTest.java (hunk 3 above)
Comment (weizhouapache) [reply plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParserTest.java:52]: done , thanks @wido

File: plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParserTest.java (hunk 2 above)
Comment (weizhouapache) [reply plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParserTest.java:76]: done

**Comment by weizhouapache** [comment]
@wido 
I have made some changes to support ceph monitors with IPv6 addresses. some test cases are added as well.

I am seriously wondering if you are using ceph with IPv6 address in your production, because I faced some blocker issues in my testing
(1) failed to copy template to ceph storage when deploy a vm.  this is fixed by https://github.com/apache/cloudstack/pull/6792/commits/4a242a76f93acff379023bd638ff442912254cc4
(2) failed to take volume snapshot, this is also fixed by commit https://github.com/apache/cloudstack/pull/6792/commits/4a242a76f93acff379023bd638ff442912254cc4
(3) failed to start vm
when I create storage pool, the ipv6 addresses must have brackets ("[" and "]"), but the brackets must be removed in vm definition xml. this is fixed by https://github.com/apache/cloudstack/pull/6792/commits/34f8732a0c30a8e890ecb9f21b20613054580691

With all commits in this PR, the RBD storage pools with single or multiple Ipv4/ipv6 monitors work well in my testing.
I will continue on testing. 
now it is ready for code reviews. @wido @DaanHoogland @rohityadavcloud @borisstoyanov

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by wido** [comment]
> @wido I have made some changes to support ceph monitors with IPv6 addresses. some test cases are added as well.
> 
> I am seriously wondering if you are using ceph with IPv6 address in your production, because I faced some blocker issues in my testing (1) failed to copy template to ceph storage when deploy a vm. this is fixed by [4a242a7](https://github.com/apache/cloudstack/commit/4a242a76f93acff379023bd638ff442912254cc4) (2) failed to take volume snapshot, this is also fixed by commit [4a242a7](https://github.com/apache/cloudstack/commit/4a242a76f93acff379023bd638ff442912254cc4) (3) failed to start vm when I create storage pool, the ipv6 addresses must have brackets ("[" and "]"), but the brackets must be removed in vm definition xml. this is fixed by [34f8732](https://github.com/apache/cloudstack/commit/34f8732a0c30a8e890ecb9f21b20613054580691)
> 
> With all commits in this PR, the RBD storage pools with single or multiple Ipv4/ipv6 monitors work well in my testing. I will continue on testing. now it is ready for code reviews. @wido @DaanHoogland @rohityadavcloud @borisstoyanov

Thank you, I'll take a look.

We use Ceph with IPv6, but we only use hostnames for the Ceph monitors. A Round Robin DNS record pointing to the 3 MONs. We do not use manual IPv6 addresses in the host field.

**Comment by weizhouapache** [comment]
> > @wido I have made some changes to support ceph monitors with IPv6 addresses. some test cases are added as well.
> > I am seriously wondering if you are using ceph with IPv6 address in your production, because I faced some blocker issues in my testing (1) failed to copy template to ceph storage when deploy a vm. this is fixed by [4a242a7](https://github.com/apache/cloudstack/commit/4a242a76f93acff379023bd638ff442912254cc4) (2) failed to take volume snapshot, this is also fixed by commit [4a242a7](https://github.com/apache/cloudstack/commit/4a242a76f93acff379023bd638ff442912254cc4) (3) failed to start vm when I create storage pool, the ipv6 addresses must have brackets ("[" and "]"), but the brackets must be removed in vm definition xml. this is fixed by [34f8732](https://github.com/apache/cloudstack/commit/34f8732a0c30a8e890ecb9f21b20613054580691)
> > With all commits in this PR, the RBD storage pools with single or multiple Ipv4/ipv6 monitors work well in my testing. I will continue on testing. now it is ready for code reviews. @wido @DaanHoogland @rohityadavcloud @borisstoyanov
> 
> Thank you, I'll take a look.
> 
> We use Ceph with IPv6, but we only use hostnames for the Ceph monitors. A Round Robin DNS record pointing to the 3 MONs. We do not use manual IPv6 addresses in the host field.

thanks @wido
I think it explains why you do not the issues. I will test with (1 or 3) host names

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test ubuntu22 kvm-ubuntu22

**Comment by weizhouapache** [comment]
when testing this PR we have found an issue that 

- centos7 kvm host can NOT connect to ceph 17 (single monitor or multiple monitors)
- centos7 kvm host can connecto to ceph 15

thanks @NuxRo

**Review by NuxRo — APPROVED** [review]
Did some testing, looks good to me!

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
@wido by the looks of it all your comments are met, right?

**Review by rohityadavcloud — APPROVED** [review]

**Comment by rohityadavcloud** [comment]
LGTM - let's merge this @DaanHoogland ?

**Comment by DaanHoogland** [comment]
We are not in "freeze" @rohityadavcloud ;)

**Comment by zap51** [comment]
@rohityadavcloud @DaanHoogland A [PR](https://github.com/ceph/ceph/pull/52595) has been created on Ceph to include this.

**Pull request merged**